src/crab/variable.hpp

```diff
@@ -39,17 +40,28 @@ class variable_t final {
     bool operator<(variable_t o) const { return _id < o._id; }
 
 
-    [[nodiscard]] std::string name() const { return names.at(_id); }
+    [[nodiscard]] std::string name() const { return names->at(_id); }
 
-    [[nodiscard]] bool is_type() const { return names.at(_id).find(".type") != std::string::npos; }
+    [[nodiscard]] bool is_type() const { return names->at(_id).find(".type") != std::string::npos; }
 
-    friend std::ostream& operator<<(std::ostream& o, variable_t v)  { return o << names.at(v._id); }
+    friend std::ostream& operator<<(std::ostream& o, variable_t v)  { return o << names->at(v._id); }
 
     // var_factory portion.
     // This singleton is eBPF-specific, to avoid lifetime issues and/or passing factory explicitly everywhere:
   private:
     static variable_t make(const std::string& name);
-    static thread_local std::vector<std::string> names;
+    static std::vector<std::string> _default_names();
+
+    /**
+     * @brief Factory to alway return the initial variable names.
+     *
+     * @tparam[in] T Should always be std::vector<std::string>.
+     */
+    template <typename T>
+    struct variable_name_factory {
+        T operator()() { return _default_names(); }
+    };
```

Comment:
I think you can just specialize the template and avoid the comment. Also, I think it is possible use a simple pointer to function as the template argument.

Reply:
This was a compromise. This was the simplest option that allows callers to use lazy_allocator with just the primary type and to have the compiler insatiate the appropriate factory for the default value. Making the factory argument a plain function pointer doesn't seem to allow the compiler to instantiate it automatically (or at least I wasn't able to figure out how to do that).